### ISSUE TITLE
[backend] llama-cpp, C++ gRPC backend

### ISSUE THREAD

**Issue by mudler** [go-skynet/LocalAI#1154]
**Is your feature request related to a problem? Please describe.**
keeping in sync the golang binding can lag behind for certain reasons related to the golang toolchain. It'd be easier to also have a c++ implementation that requires less  maitenance and so can be bumped almost automatically

**Describe the solution you'd like**
At this point we can also have a pure C++ llama-cpp gRPC server binding with just the few little things we need to add on top

**Describe alternatives you've considered**

**Additional context**
This would allow us to open up bugs upstream more easily as we have a less heavy implementation on top. We could also upstream the gRPC server

**Comment by Aisuko** [comment]
Yes, this is a good solution. Sometimes is so hard to follow the log from the golang binding packages.

**Comment by mudler** [comment]
JFYI I'm currentl playing with this

**Comment by Aisuko** [comment]
Sounds really good. And I believe that the C++ backend can be an example for other languages, like: Rust. I am planing on how to implement the Rust backend. Can I say the basic requirement of our backend is:
* Same `proto` file to keep same behaviors of API
* `async` should be as default
* Only server side gRPC services

Are these the basic requirements of our backends?

**Comment by mudler** [comment]
> Sounds really good. And I believe that the C++ backend can be an example for other languages, like: Rust. I am planing on how to implement the Rust backend. Can I say the basic requirements of our backend are:
> 
>     * Same `proto` file to keep same behaviors of API
> 
>     * `async` should be as default (Streaming response should be better)
> 
>     * Only server side gRPC services
> 
> 
> Is that right?

looks reasonable to me! I think also the golang gRPC services are good examples as well, probably C++ is going to be a bit more verbose-ish
